File: network/src/main/java/com/github/vivchar/network/GithubClient.java
package com.github.vivchar.network;

import androidx.annotation.NonNull;

import com.github.vivchar.network.models.GithubFork;
import com.github.vivchar.network.models.GithubUser;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by Vivchar Vitaly on 12.10.17.
 */

class GithubClient {

	@NonNull
	private final GithubAPI mGithubAPI;
	@NonNull
	private final EventListener mListener;

	public GithubClient(@NonNull final GithubAPI githubAPI, @NonNull final EventListener listener) {
		mGithubAPI = githubAPI;
		mListener = listener;
	}

	public void sendStargazersRequest(final int page) {
		mGithubAPI.getStargazers(page).enqueue(new Callback<List<GithubUser>>() {
			@Override
			public void onResponse(final Call<List<GithubUser>> call, final Response<List<GithubUser>> response) {
				final List<GithubUser> body = response.body();
				if (response.isSuccessful() && body != null) {
					mListener.onStargazersReceived(page, body);
				} else {
					mListener.onStargazersFailed(page);
				}
			}

			@Override
			public void onFailure(final Call<List<GithubUser>> call, final Throwable t) {
				mListener.onStargazersFailed(page);
			}
		});
	}

	public void sendForksRequest() {
		mGithubAPI.getForks().enqueue(new Callback<List<GithubFork>>() {
			@Override
			public void onResponse(final Call<List<GithubFork>> call, final Response<List<GithubFork>> response) {
				final List<GithubFork> body = response.body();
				if (response.isSuccessful() && body != null) {
					mListener.onForksReceived(body);
				} else {
					mListener.onForksFailed("Response is failed");
				}
			}

			@Override
			public void onFailure(final Call<List<GithubFork>> call, final Throwable t) {
				mListener.onForksFailed(t.getMessage());
			}
		});
	}

	public interface EventListener {
		void onStargazersReceived(int page, @NonNull List<GithubUser> stargazers);
		void onStargazersFailed(int page);
		void onForksReceived(@NonNull List<GithubFork> forks);
		void onForksFailed(@NonNull final String message);
	}
}

File: network/src/main/java/com/github/vivchar/network/GithubClient.kt
package com.github.vivchar.network

import com.github.vivchar.network.models.GithubFork
import com.github.vivchar.network.models.GithubUser
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Created by Vivchar Vitaly on 12.10.17.
 */
internal class GithubClient(private val githubAPI: GithubAPI, private val listener: EventListener) {

	fun sendStargazersRequest(page: Int) {
		githubAPI.getStargazers(page).enqueue(object : Callback<List<GithubUser>> {
			override fun onResponse(call: Call<List<GithubUser>>, response: Response<List<GithubUser>>) {
				val body = response.body()
				if (response.isSuccessful && body != null) {
					listener.onStargazersReceived(page, body)
				} else {
					listener.onStargazersFailed(page)
				}
			}

			override fun onFailure(call: Call<List<GithubUser>>, t: Throwable) {
				listener.onStargazersFailed(page)
			}
		})
	}

	fun sendForksRequest() {
		githubAPI.forks.enqueue(object : Callback<List<GithubFork>> {
			override fun onResponse(call: Call<List<GithubFork>>, response: Response<List<GithubFork>>) {
				val body = response.body()
				if (response.isSuccessful && body != null) {
					listener.onForksReceived(body)
				} else {
					listener.onForksFailed("Response is failed")
				}
			}

			override fun onFailure(call: Call<List<GithubFork>>, t: Throwable) {
				listener.onForksFailed(t.message!!)
			}
		})
	}

	interface EventListener {
		fun onStargazersReceived(page: Int, stargazers: List<GithubUser>)
		fun onStargazersFailed(page: Int)
		fun onForksReceived(forks: List<GithubFork>)
		fun onForksFailed(message: String)
	}
}